File: 2021/day2.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func part1(lines []string) {
	depth, pos := 0, 0

	for _, line := range lines {
		direction, dist := parseLine(line)

		switch direction {
		case "forward":
			pos += dist
		case "up":
			depth -= dist
		default:
			depth += dist
		}
	}

	fmt.Println(depth * pos)
}

func part2(lines []string) {
	pos, depth, aim := 0, 0, 0
	for _, line := range lines {
		direction, dist := parseLine(line)

		if direction == "forward" {
			pos += dist
			depth += dist * aim
		} else if direction == "up" {
			aim -= dist
		} else {
			aim += dist
		}
	}

	fmt.Println(depth * pos)
}

func parseLine(line string) (string, int) {
	cmd := strings.Split(line, " ")
	direction := cmd[0]
	dist, _ := strconv.Atoi(cmd[1])
	return direction, dist
}

func main() {
	data, _ := os.ReadFile("2021/input/input_2.txt")
	file_content := string(data)
	lines := strings.Split(file_content, "\n")
	lines = lines[:len(lines)-1]

	part1(lines)
	part2(lines)
}

File: 2021/day2.py
import sys
from pathlib import Path

path_root = Path(__file__).parents[1]
sys.path.append(str(path_root))

from utils.get_inputs import ProblemParser


lines = ProblemParser().load_input(2021, 2)


def parse_line(line):
    l = line.split(" ")
    direction = l[0]
    dist = int(l[1])
    return direction, dist


def part1():
    position = 0
    depth = 0

    for line in lines:
        direction, dist = parse_line(line)
        if direction == "forward":
            position += dist
        elif direction == "down":
            depth += dist
        else:
            depth -= dist

    print(position * depth)


def part2():
    position = 0
    depth = 0
    aim = 0

    for line in lines:
        direction, dist = parse_line(line)
        if direction == "forward":
            position += dist
            depth += aim * dist
        elif direction == "down":
            aim += dist
        else:
            aim -= dist

    print(position * depth)


part1()
part2()
